// @decartai/sdk/dist/index.js
import { ERROR_CODES, createInvalidApiKeyError, createInvalidBaseUrlError } from "./utils/errors.js";
import { createProcessClient } from "./process/client.js";
import { createQueueClient } from "./queue/client.js";
import { isImageModel, isRealtimeModel, isVideoModel, models } from "./shared/model.js";
import { createRealTimeClient } from "./realtime/client.js";
import { createTokensClient } from "./tokens/client.js";
import { readEnv } from "./utils/env.js";
import { createConsoleLogger, noopLogger } from "./utils/logger.js";
import { z } from "zod";

//#region src/index.ts
const proxySchema = z.union([z.string().url(), z.string().startsWith("/")]);
const decartClientOptionsSchema = z.object({
	apiKey: z.string().min(1).optional(),
	baseUrl: z.url().optional(),
	proxy: proxySchema.optional(),
	integration: z.string().optional(),
	realtimeBaseUrl: z.url().optional()
}).refine((data) => {
	const hasProxy = !!data.proxy;
	const hasApiKey = !!data.apiKey;
	return !(hasProxy && hasApiKey);
}, { message: "Cannot provide both 'proxy' and 'apiKey'. Use 'proxy' for proxy mode or 'apiKey' for direct API access." });
/**
* Create a Decart API client.
*
* @param options - Configuration options
* @param options.proxy - URL of the proxy server. When set, the client will use the proxy instead of direct API access and apiKey is not required.
* @param options.apiKey - API key for authentication.
* @param options.baseUrl - Override the default API base URL.
* @param options.realtimeBaseUrl - Override the default WebSocket base URL for realtime connections.
* @param options.integration - Optional integration identifier.
*
* @example
* ```ts
* //  (direct API access)Option 1: Explicit API key
* const client = createDecartClient({ apiKey: "your-api-key" });
*
* // Option 2: Using DECART_API_KEY environment variable
* const client = createDecartClient();
*
* // Option 3: Using proxy (client-side, no API key needed)
* const client = createDecartClient({ proxy: "https://your-server.com/api/decart" });
* ```
*/
const createDecartClient = (options = {}) => {
	const parsedOptions = decartClientOptionsSchema.safeParse(options);
	if (!parsedOptions.success) {
		const issue = parsedOptions.error.issues[0];
		if (issue.path.includes("apiKey")) throw createInvalidApiKeyError();
		if (issue.path.includes("baseUrl") || issue.path.includes("realtimeBaseUrl")) throw createInvalidBaseUrlError(options[issue.path.includes("realtimeBaseUrl") ? "realtimeBaseUrl" : "baseUrl"]);
		if (issue.path.includes("proxy")) throw createInvalidBaseUrlError(issue.path.includes("proxy") ? options.proxy : void 0);
		throw parsedOptions.error;
	}
	const isProxyMode = "proxy" in parsedOptions.data && !!parsedOptions.data.proxy;
	const apiKey = isProxyMode ? void 0 : ("apiKey" in parsedOptions.data ? parsedOptions.data.apiKey : void 0) ?? readEnv("DECART_API_KEY");
	if (!isProxyMode && !apiKey) throw createInvalidApiKeyError();
	let baseUrl;
	if (isProxyMode && "proxy" in parsedOptions.data && parsedOptions.data.proxy) baseUrl = parsedOptions.data.proxy;
	else baseUrl = parsedOptions.data.baseUrl || "https://api.decart.ai";
	const { integration } = parsedOptions.data;
	const logger = "logger" in options && options.logger ? options.logger : noopLogger;
	const telemetryEnabled = "telemetry" in options && options.telemetry === false ? false : true;
	return {
		realtime: createRealTimeClient({
			baseUrl: parsedOptions.data.realtimeBaseUrl || "wss://api3.decart.ai",
			apiKey: apiKey || "",
			integration,
			logger,
			telemetryEnabled
		}),
		process: createProcessClient({
			baseUrl,
			apiKey: apiKey || "",
			integration
		}),
		queue: createQueueClient({
			baseUrl,
			apiKey: apiKey || "",
			integration
		}),
		tokens: createTokensClient({
			baseUrl,
			apiKey: apiKey || "",
			integration
		})
	};
};

//#endregion
export { ERROR_CODES, createConsoleLogger, createDecartClient, isImageModel, isRealtimeModel, isVideoModel, models, noopLogger };
