// @decartai/sdk/dist/utils/errors.js
//#region src/utils/errors.ts
const ERROR_CODES = {
	INVALID_API_KEY: "INVALID_API_KEY",
	INVALID_BASE_URL: "INVALID_BASE_URL",
	PROCESSING_ERROR: "PROCESSING_ERROR",
	INVALID_INPUT: "INVALID_INPUT",
	INVALID_OPTIONS: "INVALID_OPTIONS",
	MODEL_NOT_FOUND: "MODEL_NOT_FOUND",
	QUEUE_SUBMIT_ERROR: "QUEUE_SUBMIT_ERROR",
	QUEUE_STATUS_ERROR: "QUEUE_STATUS_ERROR",
	QUEUE_RESULT_ERROR: "QUEUE_RESULT_ERROR",
	JOB_NOT_COMPLETED: "JOB_NOT_COMPLETED",
	TOKEN_CREATE_ERROR: "TOKEN_CREATE_ERROR",
	WEBRTC_WEBSOCKET_ERROR: "WEBRTC_WEBSOCKET_ERROR",
	WEBRTC_ICE_ERROR: "WEBRTC_ICE_ERROR",
	WEBRTC_TIMEOUT_ERROR: "WEBRTC_TIMEOUT_ERROR",
	WEBRTC_SERVER_ERROR: "WEBRTC_SERVER_ERROR",
	WEBRTC_SIGNALING_ERROR: "WEBRTC_SIGNALING_ERROR"
};
function createSDKError(code, message, data, cause) {
	return {
		code,
		message,
		data,
		cause
	};
}
function createInvalidApiKeyError() {
	return createSDKError(ERROR_CODES.INVALID_API_KEY, "Missing API key. Pass `apiKey` to createDecartClient() or set the DECART_API_KEY environment variable.");
}
function createInvalidBaseUrlError(url) {
	return createSDKError(ERROR_CODES.INVALID_BASE_URL, `Invalid base URL${url ? `: ${url}` : ""}`);
}
function createWebrtcWebsocketError(error) {
	return createSDKError(ERROR_CODES.WEBRTC_WEBSOCKET_ERROR, "WebSocket connection failed", void 0, error);
}
function createWebrtcIceError(error) {
	return createSDKError(ERROR_CODES.WEBRTC_ICE_ERROR, "ICE connection failed", void 0, error);
}
function createWebrtcTimeoutError(phase, timeoutMs, cause) {
	const hasTimeout = typeof timeoutMs === "number" && Number.isFinite(timeoutMs);
	return createSDKError(ERROR_CODES.WEBRTC_TIMEOUT_ERROR, hasTimeout ? `${phase} timed out after ${timeoutMs}ms` : `${phase} timed out`, hasTimeout ? {
		phase,
		timeoutMs
	} : { phase }, cause);
}
function createWebrtcServerError(message) {
	return createSDKError(ERROR_CODES.WEBRTC_SERVER_ERROR, message);
}
function createWebrtcSignalingError(error) {
	return createSDKError(ERROR_CODES.WEBRTC_SIGNALING_ERROR, "Signaling error", void 0, error);
}
/**
* Classify a raw WebRTC error into a specific SDK error based on its message.
*/
function classifyWebrtcError(error) {
	const msg = error.message.toLowerCase();
	if (error.source === "server") return createWebrtcServerError(error.message);
	if (msg.includes("websocket")) return createWebrtcWebsocketError(error);
	if (msg.includes("ice connection failed")) return createWebrtcIceError(error);
	if (msg.includes("timeout") || msg.includes("timed out")) {
		const timeoutMatch = msg.match(/(\d+)\s*ms/);
		return createWebrtcTimeoutError("connection", timeoutMatch ? Number.parseInt(timeoutMatch[1], 10) : void 0, error);
	}
	return createWebrtcSignalingError(error);
}
function createInvalidInputError(message) {
	return createSDKError(ERROR_CODES.INVALID_INPUT, message);
}
function createModelNotFoundError(model) {
	return createSDKError(ERROR_CODES.MODEL_NOT_FOUND, `Model ${model} not found`);
}
function createQueueSubmitError(message, status) {
	return createSDKError(ERROR_CODES.QUEUE_SUBMIT_ERROR, message, { status });
}
function createQueueStatusError(message, status) {
	return createSDKError(ERROR_CODES.QUEUE_STATUS_ERROR, message, { status });
}
function createQueueResultError(message, status) {
	return createSDKError(ERROR_CODES.QUEUE_RESULT_ERROR, message, { status });
}

//#endregion
export { ERROR_CODES, classifyWebrtcError, createInvalidApiKeyError, createInvalidBaseUrlError, createInvalidInputError, createModelNotFoundError, createQueueResultError, createQueueStatusError, createQueueSubmitError, createSDKError };
