// @decartai/sdk/dist/realtime/audio-stream-manager.js
//#region src/realtime/audio-stream-manager.ts
/**
* Manages an audio stream for live_avatar mode.
* Creates a continuous audio stream that outputs silence by default,
* and allows playing audio files through the stream.
*/
var AudioStreamManager = class {
	audioContext;
	destination;
	silenceOscillator;
	silenceGain;
	currentSource = null;
	_isPlaying = false;
	constructor() {
		this.audioContext = new AudioContext();
		this.destination = this.audioContext.createMediaStreamDestination();
		this.silenceOscillator = this.audioContext.createOscillator();
		this.silenceGain = this.audioContext.createGain();
		this.silenceGain.gain.value = 0;
		this.silenceOscillator.connect(this.silenceGain);
		this.silenceGain.connect(this.destination);
		this.silenceOscillator.start();
	}
	/**
	* Get the MediaStream to pass to WebRTC.
	* This stream outputs silence when no audio is playing.
	*/
	getStream() {
		return this.destination.stream;
	}
	/**
	* Check if audio is currently playing.
	*/
	isPlaying() {
		return this._isPlaying;
	}
	/**
	* Play audio through the stream.
	* When the audio ends, the stream automatically reverts to silence.
	* @param audio - Audio data as Blob, File, or ArrayBuffer
	* @returns Promise that resolves when audio finishes playing
	*/
	async playAudio(audio) {
		if (this.audioContext.state === "suspended") await this.audioContext.resume();
		if (this._isPlaying) this.stopAudio();
		let arrayBuffer;
		if (audio instanceof ArrayBuffer) arrayBuffer = audio;
		else arrayBuffer = await audio.arrayBuffer();
		const audioBuffer = await this.audioContext.decodeAudioData(arrayBuffer);
		const source = this.audioContext.createBufferSource();
		source.buffer = audioBuffer;
		source.connect(this.destination);
		this.currentSource = source;
		this._isPlaying = true;
		return new Promise((resolve) => {
			source.onended = () => {
				this._isPlaying = false;
				this.currentSource = null;
				resolve();
			};
			source.start();
		});
	}
	/**
	* Stop currently playing audio immediately.
	* The stream will revert to silence.
	*/
	stopAudio() {
		if (this.currentSource) {
			try {
				this.currentSource.stop();
			} catch {}
			this.currentSource = null;
		}
		this._isPlaying = false;
	}
	/**
	* Clean up resources.
	*/
	cleanup() {
		this.stopAudio();
		try {
			this.silenceOscillator.stop();
		} catch {}
		this.audioContext.close().catch(() => {});
	}
};

//#endregion
export { AudioStreamManager };
