// @decartai/sdk/dist/realtime/client.js
import { classifyWebrtcError } from "../utils/errors.js";
import { modelDefinitionSchema } from "../shared/model.js";
import { modelStateSchema } from "../shared/types.js";
import { AudioStreamManager } from "./audio-stream-manager.js";
import { createEventBuffer } from "./event-buffer.js";
import { realtimeMethods } from "./methods.js";
import { decodeSubscribeToken, encodeSubscribeToken } from "./subscribe-client.js";
import { NullTelemetryReporter, TelemetryReporter } from "./telemetry-reporter.js";
import { WebRTCManager } from "./webrtc-manager.js";
import { WebRTCStatsCollector } from "./webrtc-stats.js";
import { z } from "zod";

//#region src/realtime/client.ts
async function blobToBase64(blob) {
	return new Promise((resolve, reject) => {
		const reader = new FileReader();
		reader.onloadend = () => {
			const result = reader.result;
			if (typeof result !== "string") {
				reject(/* @__PURE__ */ new Error("FileReader did not return a string"));
				return;
			}
			const base64 = result.split(",")[1];
			if (!base64) {
				reject(/* @__PURE__ */ new Error("Invalid data URL format"));
				return;
			}
			resolve(base64);
		};
		reader.onerror = reject;
		reader.readAsDataURL(blob);
	});
}
async function imageToBase64(image) {
	if (typeof image === "string") {
		let url = null;
		try {
			url = new URL(image);
		} catch {}
		if (url?.protocol === "data:") {
			const [, base64] = image.split(",", 2);
			if (!base64) throw new Error("Invalid data URL image");
			return base64;
		}
		if (url?.protocol === "http:" || url?.protocol === "https:") {
			const response = await fetch(image);
			if (!response.ok) throw new Error(`Failed to fetch image: ${response.status} ${response.statusText}`);
			return blobToBase64(await response.blob());
		}
		return image;
	}
	return blobToBase64(image);
}
const realTimeClientInitialStateSchema = modelStateSchema;
const createAsyncFunctionSchema = (schema) => z.custom((fn) => schema.implementAsync(fn));
const realTimeClientConnectOptionsSchema = z.object({
	model: modelDefinitionSchema,
	onRemoteStream: z.custom((val) => typeof val === "function", { message: "onRemoteStream must be a function" }),
	initialState: realTimeClientInitialStateSchema.optional(),
	customizeOffer: createAsyncFunctionSchema(z.function()).optional()
});
const createRealTimeClient = (opts) => {
	const { baseUrl, apiKey, integration, logger } = opts;
	const connect = async (stream, options) => {
		const parsedOptions = realTimeClientConnectOptionsSchema.safeParse(options);
		if (!parsedOptions.success) throw parsedOptions.error;
		const isAvatarLive = options.model.name === "live_avatar" || options.model.name === "live-avatar";
		const { onRemoteStream, initialState } = parsedOptions.data;
		let audioStreamManager;
		let inputStream;
		if (isAvatarLive && !stream) {
			audioStreamManager = new AudioStreamManager();
			inputStream = audioStreamManager.getStream();
		} else inputStream = stream ?? new MediaStream();
		let webrtcManager;
		let telemetryReporter = new NullTelemetryReporter();
		let handleConnectionStateChange = null;
		try {
			const initialImage = initialState?.image ? await imageToBase64(initialState.image) : void 0;
			const initialPrompt = initialState?.prompt ? {
				text: initialState.prompt.text,
				enhance: initialState.prompt.enhance
			} : void 0;
			const url = `${baseUrl}${options.model.urlPath}`;
			const { emitter: eventEmitter, emitOrBuffer, flush, stop } = createEventBuffer();
			webrtcManager = new WebRTCManager({
				webrtcUrl: `${url}?api_key=${encodeURIComponent(apiKey)}&model=${encodeURIComponent(options.model.name)}`,
				integration,
				logger,
				onDiagnostic: (name, data) => {
					emitOrBuffer("diagnostic", {
						name,
						data
					});
					addTelemetryDiagnostic(name, data);
				},
				onRemoteStream,
				onConnectionStateChange: (state) => {
					emitOrBuffer("connectionChange", state);
					handleConnectionStateChange?.(state);
				},
				onError: (error) => {
					logger.error("WebRTC error", { error: error.message });
					emitOrBuffer("error", classifyWebrtcError(error));
				},
				customizeOffer: options.customizeOffer,
				vp8MinBitrate: 300,
				vp8StartBitrate: 600,
				modelName: options.model.name,
				initialImage,
				initialPrompt
			});
			const manager = webrtcManager;
			let sessionId = null;
			let subscribeToken = null;
			const pendingTelemetryDiagnostics = [];
			let telemetryReporterReady = false;
			const addTelemetryDiagnostic = (name, data, timestamp = Date.now()) => {
				if (!opts.telemetryEnabled) return;
				if (!telemetryReporterReady) {
					pendingTelemetryDiagnostics.push({
						name,
						data,
						timestamp
					});
					return;
				}
				telemetryReporter.addDiagnostic({
					name,
					data,
					timestamp
				});
			};
			const sessionIdListener = (msg) => {
				subscribeToken = encodeSubscribeToken(msg.session_id, msg.server_ip, msg.server_port);
				sessionId = msg.session_id;
				if (opts.telemetryEnabled) {
					if (telemetryReporterReady) telemetryReporter.stop();
					const reporter = new TelemetryReporter({
						apiKey,
						sessionId: msg.session_id,
						model: options.model.name,
						integration,
						logger
					});
					reporter.start();
					telemetryReporter = reporter;
					telemetryReporterReady = true;
					for (const diagnostic of pendingTelemetryDiagnostics) telemetryReporter.addDiagnostic(diagnostic);
					pendingTelemetryDiagnostics.length = 0;
				}
			};
			manager.getWebsocketMessageEmitter().on("sessionId", sessionIdListener);
			const tickListener = (msg) => {
				emitOrBuffer("generationTick", { seconds: msg.seconds });
			};
			manager.getWebsocketMessageEmitter().on("generationTick", tickListener);
			await manager.connect(inputStream);
			const methods = realtimeMethods(manager, imageToBase64);
			let statsCollector = null;
			let statsCollectorPeerConnection = null;
			const STALL_FPS_THRESHOLD = .5;
			let videoStalled = false;
			let stallStartMs = 0;
			const startStatsCollection = () => {
				statsCollector?.stop();
				videoStalled = false;
				stallStartMs = 0;
				statsCollector = new WebRTCStatsCollector();
				const pc = manager.getPeerConnection();
				statsCollectorPeerConnection = pc;
				if (pc) statsCollector.start(pc, (stats) => {
					emitOrBuffer("stats", stats);
					telemetryReporter.addStats(stats);
					const fps = stats.video?.framesPerSecond ?? 0;
					if (!videoStalled && stats.video && fps < STALL_FPS_THRESHOLD) {
						videoStalled = true;
						stallStartMs = Date.now();
						emitOrBuffer("diagnostic", {
							name: "videoStall",
							data: {
								stalled: true,
								durationMs: 0
							}
						});
						addTelemetryDiagnostic("videoStall", {
							stalled: true,
							durationMs: 0
						}, stallStartMs);
					} else if (videoStalled && fps >= STALL_FPS_THRESHOLD) {
						const durationMs = Date.now() - stallStartMs;
						videoStalled = false;
						emitOrBuffer("diagnostic", {
							name: "videoStall",
							data: {
								stalled: false,
								durationMs
							}
						});
						addTelemetryDiagnostic("videoStall", {
							stalled: false,
							durationMs
						});
					}
				});
				return () => {
					statsCollector?.stop();
					statsCollector = null;
					statsCollectorPeerConnection = null;
				};
			};
			handleConnectionStateChange = (state) => {
				if (!opts.telemetryEnabled) return;
				if (state !== "connected" && state !== "generating") return;
				const peerConnection = manager.getPeerConnection();
				if (!peerConnection || peerConnection === statsCollectorPeerConnection) return;
				startStatsCollection();
			};
			if (opts.telemetryEnabled) startStatsCollection();
			const client = {
				set: methods.set,
				setPrompt: methods.setPrompt,
				isConnected: () => manager.isConnected(),
				getConnectionState: () => manager.getConnectionState(),
				disconnect: () => {
					statsCollector?.stop();
					telemetryReporter.stop();
					stop();
					manager.cleanup();
					audioStreamManager?.cleanup();
				},
				on: eventEmitter.on,
				off: eventEmitter.off,
				get sessionId() {
					return sessionId;
				},
				get subscribeToken() {
					return subscribeToken;
				},
				setImage: async (image, options) => {
					if (image === null) return manager.setImage(null, options);
					const base64 = await imageToBase64(image);
					return manager.setImage(base64, options);
				}
			};
			if (isAvatarLive && audioStreamManager) {
				const manager = audioStreamManager;
				client.playAudio = (audio) => manager.playAudio(audio);
			}
			flush();
			return client;
		} catch (error) {
			telemetryReporter.stop();
			webrtcManager?.cleanup();
			audioStreamManager?.cleanup();
			throw error;
		}
	};
	const subscribe = async (options) => {
		const { sid, ip, port } = decodeSubscribeToken(options.token);
		const subscribeUrl = `${baseUrl}/subscribe/${encodeURIComponent(sid)}?IP=${encodeURIComponent(ip)}&port=${encodeURIComponent(port)}&api_key=${encodeURIComponent(apiKey)}`;
		const { emitter: eventEmitter, emitOrBuffer, flush, stop } = createEventBuffer();
		let webrtcManager;
		try {
			webrtcManager = new WebRTCManager({
				webrtcUrl: subscribeUrl,
				integration,
				logger,
				onDiagnostic: (name, data) => {
					emitOrBuffer("diagnostic", {
						name,
						data
					});
				},
				onRemoteStream: options.onRemoteStream,
				onConnectionStateChange: (state) => {
					emitOrBuffer("connectionChange", state);
				},
				onError: (error) => {
					logger.error("WebRTC subscribe error", { error: error.message });
					emitOrBuffer("error", classifyWebrtcError(error));
				}
			});
			const manager = webrtcManager;
			await manager.connect(null);
			const client = {
				isConnected: () => manager.isConnected(),
				getConnectionState: () => manager.getConnectionState(),
				disconnect: () => {
					stop();
					manager.cleanup();
				},
				on: eventEmitter.on,
				off: eventEmitter.off
			};
			flush();
			return client;
		} catch (error) {
			webrtcManager?.cleanup();
			throw error;
		}
	};
	return {
		connect,
		subscribe
	};
};

//#endregion
export { createRealTimeClient };
