// @decartai/sdk/dist/realtime/methods.js
import { z } from "zod";

//#region src/realtime/methods.ts
const PROMPT_TIMEOUT_MS = 15 * 1e3;
const UPDATE_TIMEOUT_MS = 30 * 1e3;
const setInputSchema = z.object({
	prompt: z.string().min(1).optional(),
	enhance: z.boolean().optional().default(true),
	image: z.union([
		z.instanceof(Blob),
		z.instanceof(File),
		z.string(),
		z.null()
	]).optional()
}).refine((data) => data.prompt !== void 0 || data.image !== void 0, { message: "At least one of 'prompt' or 'image' must be provided" });
const setPromptInputSchema = z.object({
	prompt: z.string().min(1),
	enhance: z.boolean().optional().default(true)
});
const realtimeMethods = (webrtcManager, imageToBase64) => {
	const assertConnected = () => {
		const state = webrtcManager.getConnectionState();
		if (state !== "connected" && state !== "generating") throw new Error(`Cannot send message: connection is ${state}`);
	};
	const set = async (input) => {
		assertConnected();
		const parsed = setInputSchema.safeParse(input);
		if (!parsed.success) throw parsed.error;
		const { prompt, enhance, image } = parsed.data;
		let imageBase64 = null;
		if (image !== void 0 && image !== null) imageBase64 = await imageToBase64(image);
		await webrtcManager.setImage(imageBase64, {
			prompt,
			enhance,
			timeout: UPDATE_TIMEOUT_MS
		});
	};
	const setPrompt = async (prompt, { enhance } = {}) => {
		assertConnected();
		const parsedInput = setPromptInputSchema.safeParse({
			prompt,
			enhance
		});
		if (!parsedInput.success) throw parsedInput.error;
		const emitter = webrtcManager.getWebsocketMessageEmitter();
		let promptAckListener;
		let timeoutId;
		try {
			const ackPromise = new Promise((resolve, reject) => {
				promptAckListener = (promptAckMessage) => {
					if (promptAckMessage.prompt === parsedInput.data.prompt) if (promptAckMessage.success) resolve();
					else reject(new Error(promptAckMessage.error ?? "Failed to send prompt"));
				};
				emitter.on("promptAck", promptAckListener);
			});
			if (!webrtcManager.sendMessage({
				type: "prompt",
				prompt: parsedInput.data.prompt,
				enhance_prompt: parsedInput.data.enhance
			})) throw new Error("WebSocket is not open");
			const timeoutPromise = new Promise((_, reject) => {
				timeoutId = setTimeout(() => reject(/* @__PURE__ */ new Error("Prompt timed out")), PROMPT_TIMEOUT_MS);
			});
			await Promise.race([ackPromise, timeoutPromise]);
		} finally {
			if (promptAckListener) emitter.off("promptAck", promptAckListener);
			if (timeoutId) clearTimeout(timeoutId);
		}
	};
	return {
		set,
		setPrompt
	};
};

//#endregion
export { realtimeMethods };
