// @decartai/sdk/dist/realtime/telemetry-reporter.js
import { VERSION } from "../version.js";
import { buildAuthHeaders } from "../shared/request.js";

//#region src/realtime/telemetry-reporter.ts
const DEFAULT_REPORT_INTERVAL_MS = 1e4;
const TELEMETRY_URL = "https://platform.decart.ai/api/v1/telemetry";
/**
* Maximum number of items per array (stats / diagnostics) in a single report.
* Matches the backend Zod schema which enforces `z.array().max(120)`.
*/
const MAX_ITEMS_PER_REPORT = 120;
/** No-op reporter that silently discards all data. Used when telemetry is disabled. */
var NullTelemetryReporter = class {
	start() {}
	addStats() {}
	addDiagnostic() {}
	flush() {}
	stop() {}
};
var TelemetryReporter = class {
	apiKey;
	sessionId;
	model;
	integration;
	logger;
	reportIntervalMs;
	intervalId = null;
	statsBuffer = [];
	diagnosticsBuffer = [];
	constructor(options) {
		this.apiKey = options.apiKey;
		this.sessionId = options.sessionId;
		this.model = options.model;
		this.integration = options.integration;
		this.logger = options.logger;
		this.reportIntervalMs = options.reportIntervalMs ?? DEFAULT_REPORT_INTERVAL_MS;
	}
	/** Start the periodic reporting timer. */
	start() {
		if (this.intervalId !== null) return;
		this.intervalId = setInterval(() => this.flush(), this.reportIntervalMs);
	}
	/** Add a stats snapshot to the buffer. */
	addStats(stats) {
		this.statsBuffer.push(stats);
	}
	/** Add a diagnostic event to the buffer. */
	addDiagnostic(event) {
		this.diagnosticsBuffer.push(event);
	}
	/** Flush buffered data immediately. */
	flush() {
		this.sendReport(false);
	}
	/** Stop the reporter and send a final report with keepalive. */
	stop() {
		if (this.intervalId !== null) {
			clearInterval(this.intervalId);
			this.intervalId = null;
		}
		this.sendReport(true);
	}
	/**
	* Build a single chunk from the front of the buffers, respecting MAX_ITEMS_PER_REPORT.
	* Returns null when both buffers are empty.
	*/
	createReportChunk() {
		if (this.statsBuffer.length === 0 && this.diagnosticsBuffer.length === 0) return null;
		const tags = {
			session_id: this.sessionId,
			sdk_version: VERSION,
			...this.model ? { model: this.model } : {},
			...this.integration ? { integration: this.integration } : {}
		};
		return {
			sessionId: this.sessionId,
			timestamp: Date.now(),
			sdkVersion: VERSION,
			...this.model ? { model: this.model } : {},
			tags,
			stats: this.statsBuffer.splice(0, MAX_ITEMS_PER_REPORT),
			diagnostics: this.diagnosticsBuffer.splice(0, MAX_ITEMS_PER_REPORT)
		};
	}
	sendReport(keepalive) {
		if (this.statsBuffer.length === 0 && this.diagnosticsBuffer.length === 0) return;
		try {
			const commonHeaders = {
				...buildAuthHeaders({
					apiKey: this.apiKey,
					integration: this.integration
				}),
				"Content-Type": "application/json"
			};
			let chunk = this.createReportChunk();
			while (chunk !== null) {
				const isLast = this.statsBuffer.length === 0 && this.diagnosticsBuffer.length === 0;
				fetch(TELEMETRY_URL, {
					method: "POST",
					headers: commonHeaders,
					body: JSON.stringify(chunk),
					keepalive: keepalive && isLast
				}).then((response) => {
					if (!response.ok) this.logger.warn("Telemetry report rejected", {
						status: response.status,
						statusText: response.statusText
					});
				}).catch((error) => {
					this.logger.debug("Telemetry report failed", { error: String(error) });
				});
				chunk = this.createReportChunk();
			}
		} catch (error) {
			this.logger.debug("Telemetry report failed", { error: String(error) });
		}
	}
};

//#endregion
export { NullTelemetryReporter, TelemetryReporter };
