// @decartai/sdk/dist/realtime/webrtc-manager.js
import { WebRTCConnection } from "./webrtc-connection.js";
import pRetry, { AbortError } from "p-retry";

//#region src/realtime/webrtc-manager.ts
const PERMANENT_ERRORS = [
	"permission denied",
	"not allowed",
	"invalid session",
	"401",
	"invalid api key",
	"unauthorized"
];
const CONNECTION_TIMEOUT = 6e4 * 5;
const RETRY_OPTIONS = {
	retries: 5,
	factor: 2,
	minTimeout: 1e3,
	maxTimeout: 1e4
};
var WebRTCManager = class {
	connection;
	config;
	logger;
	localStream = null;
	subscribeMode = false;
	managerState = "disconnected";
	hasConnected = false;
	isReconnecting = false;
	intentionalDisconnect = false;
	reconnectGeneration = 0;
	constructor(config) {
		this.config = config;
		this.logger = config.logger ?? {
			debug() {},
			info() {},
			warn() {},
			error() {}
		};
		this.connection = new WebRTCConnection({
			onRemoteStream: config.onRemoteStream,
			onStateChange: (state) => this.handleConnectionStateChange(state),
			onError: config.onError,
			customizeOffer: config.customizeOffer,
			vp8MinBitrate: config.vp8MinBitrate,
			vp8StartBitrate: config.vp8StartBitrate,
			modelName: config.modelName,
			initialImage: config.initialImage,
			initialPrompt: config.initialPrompt,
			logger: this.logger,
			onDiagnostic: config.onDiagnostic
		});
	}
	emitState(state) {
		if (this.managerState !== state) {
			this.managerState = state;
			if (state === "connected" || state === "generating") this.hasConnected = true;
			this.config.onConnectionStateChange?.(state);
		}
	}
	handleConnectionStateChange(state) {
		if (this.intentionalDisconnect) {
			this.emitState("disconnected");
			return;
		}
		if (this.isReconnecting) {
			if (state === "connected" || state === "generating") {
				this.isReconnecting = false;
				this.emitState(state);
			}
			return;
		}
		if (state === "disconnected" && !this.intentionalDisconnect && this.hasConnected) {
			this.reconnect();
			return;
		}
		this.emitState(state);
	}
	async reconnect() {
		if (this.isReconnecting || this.intentionalDisconnect) return;
		if (!this.subscribeMode && !this.localStream) return;
		const reconnectGeneration = ++this.reconnectGeneration;
		this.isReconnecting = true;
		this.emitState("reconnecting");
		const reconnectStart = performance.now();
		try {
			let attemptCount = 0;
			await pRetry(async () => {
				attemptCount++;
				if (this.intentionalDisconnect || reconnectGeneration !== this.reconnectGeneration) throw new AbortError("Reconnect cancelled");
				if (!this.subscribeMode && !this.localStream) throw new AbortError("Reconnect cancelled: no local stream");
				this.connection.cleanup();
				await this.connection.connect(this.config.webrtcUrl, this.localStream, CONNECTION_TIMEOUT, this.config.integration);
				if (this.intentionalDisconnect || reconnectGeneration !== this.reconnectGeneration) {
					this.connection.cleanup();
					throw new AbortError("Reconnect cancelled");
				}
			}, {
				...RETRY_OPTIONS,
				onFailedAttempt: (error) => {
					if (this.intentionalDisconnect || reconnectGeneration !== this.reconnectGeneration) return;
					this.logger.warn("Reconnect attempt failed", {
						error: error.message,
						attempt: error.attemptNumber
					});
					this.config.onDiagnostic?.("reconnect", {
						attempt: error.attemptNumber,
						maxAttempts: RETRY_OPTIONS.retries + 1,
						durationMs: performance.now() - reconnectStart,
						success: false,
						error: error.message
					});
					this.connection.cleanup();
				},
				shouldRetry: (error) => {
					if (this.intentionalDisconnect || reconnectGeneration !== this.reconnectGeneration) return false;
					const msg = error.message.toLowerCase();
					return !PERMANENT_ERRORS.some((err) => msg.includes(err));
				}
			});
			this.config.onDiagnostic?.("reconnect", {
				attempt: attemptCount,
				maxAttempts: RETRY_OPTIONS.retries + 1,
				durationMs: performance.now() - reconnectStart,
				success: true
			});
		} catch (error) {
			this.isReconnecting = false;
			if (this.intentionalDisconnect || reconnectGeneration !== this.reconnectGeneration) return;
			this.emitState("disconnected");
			this.config.onError?.(error instanceof Error ? error : new Error(String(error)));
		}
	}
	async connect(localStream) {
		this.localStream = localStream;
		this.subscribeMode = localStream === null;
		this.intentionalDisconnect = false;
		this.hasConnected = false;
		this.isReconnecting = false;
		this.reconnectGeneration += 1;
		this.emitState("connecting");
		return pRetry(async () => {
			if (this.intentionalDisconnect) throw new AbortError("Connect cancelled");
			await this.connection.connect(this.config.webrtcUrl, localStream, CONNECTION_TIMEOUT, this.config.integration);
			return true;
		}, {
			...RETRY_OPTIONS,
			onFailedAttempt: (error) => {
				this.logger.warn("Connection attempt failed", {
					error: error.message,
					attempt: error.attemptNumber
				});
				this.connection.cleanup();
			},
			shouldRetry: (error) => {
				if (this.intentionalDisconnect) return false;
				const msg = error.message.toLowerCase();
				return !PERMANENT_ERRORS.some((err) => msg.includes(err));
			}
		});
	}
	sendMessage(message) {
		return this.connection.send(message);
	}
	cleanup() {
		this.intentionalDisconnect = true;
		this.isReconnecting = false;
		this.reconnectGeneration += 1;
		this.connection.cleanup();
		this.localStream = null;
		this.emitState("disconnected");
	}
	isConnected() {
		return this.managerState === "connected" || this.managerState === "generating";
	}
	getConnectionState() {
		return this.managerState;
	}
	getPeerConnection() {
		return this.connection.getPeerConnection();
	}
	getWebsocketMessageEmitter() {
		return this.connection.websocketMessagesEmitter;
	}
	setImage(imageBase64, options) {
		return this.connection.setImageBase64(imageBase64, options);
	}
};

//#endregion
export { WebRTCManager };
