// @decartai/sdk/dist/realtime/webrtc-stats.js
//#region src/realtime/webrtc-stats.ts
const DEFAULT_INTERVAL_MS = 1e3;
const MIN_INTERVAL_MS = 500;
var WebRTCStatsCollector = class {
	pc = null;
	intervalId = null;
	prevBytesVideo = 0;
	prevBytesAudio = 0;
	prevBytesSentVideo = 0;
	prevTimestamp = 0;
	prevPacketsLostVideo = 0;
	prevFramesDropped = 0;
	prevFreezeCount = 0;
	prevFreezeDuration = 0;
	prevPacketsLostAudio = 0;
	onStats = null;
	intervalMs;
	constructor(options = {}) {
		this.intervalMs = Math.max(options.intervalMs ?? DEFAULT_INTERVAL_MS, MIN_INTERVAL_MS);
	}
	/** Attach to a peer connection and start polling. */
	start(pc, onStats) {
		this.stop();
		this.pc = pc;
		this.onStats = onStats;
		this.prevBytesVideo = 0;
		this.prevBytesAudio = 0;
		this.prevBytesSentVideo = 0;
		this.prevTimestamp = 0;
		this.prevPacketsLostVideo = 0;
		this.prevFramesDropped = 0;
		this.prevFreezeCount = 0;
		this.prevFreezeDuration = 0;
		this.prevPacketsLostAudio = 0;
		this.intervalId = setInterval(() => this.collect(), this.intervalMs);
	}
	/** Stop polling and release resources. */
	stop() {
		if (this.intervalId !== null) {
			clearInterval(this.intervalId);
			this.intervalId = null;
		}
		this.pc = null;
		this.onStats = null;
	}
	isRunning() {
		return this.intervalId !== null;
	}
	async collect() {
		if (!this.pc || !this.onStats) return;
		try {
			const rawStats = await this.pc.getStats();
			const stats = this.parse(rawStats);
			this.onStats(stats);
		} catch {
			this.stop();
		}
	}
	parse(rawStats) {
		const now = performance.now();
		const elapsed = this.prevTimestamp > 0 ? (now - this.prevTimestamp) / 1e3 : 0;
		let video = null;
		let audio = null;
		let outboundVideo = null;
		const connection = {
			currentRoundTripTime: null,
			availableOutgoingBitrate: null
		};
		rawStats.forEach((report) => {
			if (report.type === "inbound-rtp" && report.kind === "video") {
				const bytesReceived = report.bytesReceived ?? 0;
				const bitrate = elapsed > 0 ? (bytesReceived - this.prevBytesVideo) * 8 / elapsed : 0;
				this.prevBytesVideo = bytesReceived;
				const r = report;
				const packetsLost = r.packetsLost ?? 0;
				const framesDropped = r.framesDropped ?? 0;
				const freezeCount = r.freezeCount ?? 0;
				const freezeDuration = r.totalFreezesDuration ?? 0;
				video = {
					framesDecoded: r.framesDecoded ?? 0,
					framesDropped,
					framesPerSecond: r.framesPerSecond ?? 0,
					frameWidth: r.frameWidth ?? 0,
					frameHeight: r.frameHeight ?? 0,
					bytesReceived,
					packetsReceived: r.packetsReceived ?? 0,
					packetsLost,
					jitter: r.jitter ?? 0,
					bitrate: Math.round(bitrate),
					freezeCount,
					totalFreezesDuration: freezeDuration,
					packetsLostDelta: Math.max(0, packetsLost - this.prevPacketsLostVideo),
					framesDroppedDelta: Math.max(0, framesDropped - this.prevFramesDropped),
					freezeCountDelta: Math.max(0, freezeCount - this.prevFreezeCount),
					freezeDurationDelta: Math.max(0, freezeDuration - this.prevFreezeDuration)
				};
				this.prevPacketsLostVideo = packetsLost;
				this.prevFramesDropped = framesDropped;
				this.prevFreezeCount = freezeCount;
				this.prevFreezeDuration = freezeDuration;
			}
			if (report.type === "outbound-rtp" && report.kind === "video") {
				const r = report;
				const bytesSent = r.bytesSent ?? 0;
				const outBitrate = elapsed > 0 ? (bytesSent - this.prevBytesSentVideo) * 8 / elapsed : 0;
				this.prevBytesSentVideo = bytesSent;
				outboundVideo = {
					qualityLimitationReason: r.qualityLimitationReason ?? "none",
					qualityLimitationDurations: r.qualityLimitationDurations ?? {},
					bytesSent,
					packetsSent: r.packetsSent ?? 0,
					framesPerSecond: r.framesPerSecond ?? 0,
					frameWidth: r.frameWidth ?? 0,
					frameHeight: r.frameHeight ?? 0,
					bitrate: Math.round(outBitrate)
				};
			}
			if (report.type === "inbound-rtp" && report.kind === "audio") {
				const bytesReceived = report.bytesReceived ?? 0;
				const bitrate = elapsed > 0 ? (bytesReceived - this.prevBytesAudio) * 8 / elapsed : 0;
				this.prevBytesAudio = bytesReceived;
				const r = report;
				const audioPacketsLost = r.packetsLost ?? 0;
				audio = {
					bytesReceived,
					packetsReceived: r.packetsReceived ?? 0,
					packetsLost: audioPacketsLost,
					jitter: r.jitter ?? 0,
					bitrate: Math.round(bitrate),
					packetsLostDelta: Math.max(0, audioPacketsLost - this.prevPacketsLostAudio)
				};
				this.prevPacketsLostAudio = audioPacketsLost;
			}
			if (report.type === "candidate-pair") {
				const r = report;
				if (r.state === "succeeded") {
					connection.currentRoundTripTime = r.currentRoundTripTime ?? null;
					connection.availableOutgoingBitrate = r.availableOutgoingBitrate ?? null;
				}
			}
		});
		this.prevTimestamp = now;
		return {
			timestamp: Date.now(),
			video,
			audio,
			outboundVideo,
			connection
		};
	}
};

//#endregion
export { WebRTCStatsCollector };
